python-semantic-release/python-semantic-release
# fix(algorithm): enable maintenance prereleases

## Purpose

- Refactors the version incrementation algorithm to handle prereleases & official releases when supporting more than the latest major release

- Resolves: #861

## Rationale

This change implies that where ever you are releasing from within the repository, the only tags you know about are the ones in your direct history line.  This means that prereleases of older versions can be created which were not possible as documented in #861.  Official releases were possible of older versions but this maintains that implementation.

## How you tested

I added 2 more repo rebuild tests that demonstrate and validate the use of PSR for a dual version support environment of both official releases and prerelease variants of the last major version (after the latest major version has been released).

Note that the rebuild repos will need to demonstrate going back to the latest version and releasing a new release but currently the test infrastructure does not support building a changelog relevant to its git history.  

We also could extend a dual version environment to GitHub Flow and Git Flow strategies but I did not take the additional time to make that implementation as I don't think it added a lot of value.

## How to verify

1. Check out this pr
2. Reset 1 commit
3. Run `pytest --comprehensive`, see failures:
    - `test_trunk_repo_rebuild_dual_version_spt_official_releases_only`
    - `test_trunk_repo_rebuild_dual_version_spt_w_official_n_prereleases`
    - `test_increment_version_no_major_on_zero`
    - `test_increment_version_invalid_operation`
4. Update the repo to this PR head
5. Run `pytest --comprehensive` and see failures are resolved
